### PR TITLE
Updated graphics.

### PR DESCRIPTION
**Pull request by johandegn**
I added a bunch of new 'simple to use' methods to draw other predefined or custom chapes.
I also added som simple methods to control the overall rotation and translation of the drawings.
All is documented.

### CODE REVIEW

File: graphics.ts [hunk 1]
```diff
@@ -4,6 +4,8 @@
  */
 class Graphics {
     ctx: CanvasRenderingContext2D;
```
Comment (anbclausen) [inline graphics.ts]: please be consistent with privates, either have them or dont :)

File: graphics.ts [hunk 2]
```diff
@@ -78,6 +82,76 @@ class Graphics {
         this.ctx.font = s;
     }
 
+    /**
+     * Adds to the current translation of all drawn shapes.
```
Comment (anbclausen) [inline graphics.ts]: to all drawn shapes*

File: graphics.ts [hunk 3]
```diff
@@ -78,6 +82,76 @@ class Graphics {
...
+     */
+    resetTranslation() {
+        this.ctx.translate(-this.translation[0], -this.translation[1]);
+        this.translation = [0, 0];
+    }
+
+    /**
+     * Returns the translation coordinates as a tuple of numbers.
+     */
+    getTranslation(): [number, number] {
```
Comment (anbclausen) [inline graphics.ts]: I think adding return types is a good idea, but please be consistent :)

Comment (anbclausen) [reply graphics.ts]: (put it on everything or nothing)

File: graphics.ts [hunk 4]
```diff
@@ -97,8 +171,159 @@ class Graphics {
...
+
+    /**
+     * Fills a polygon at the given coorinates with the radius and number of sides specified.
+     * @param x The x coordinate for the center of the polygon.
+     * @param y The y coordinate for the center of the polygon.
+     * @param r The radius of the polygon.
+     * @param n The number of sides in the polygon
+     * @param v The rotation of the polygon in radians. Default is 0.
+     */
+    fillPolygon(x: number, y: number, r: number, n: number, v: number = 0) {
```
Comment (anbclausen) [inline graphics.ts]: I don't like default values and optionals, as it creates potential confusion for the user of the engine :)

File: graphics.ts [hunk 5]
```diff
@@ -97,8 +171,159 @@ class Graphics {
...
+     * @param ea The end angle for the arc, in radians.
+     */
+    drawArc(x: number, y: number, r:number, sa:number, ea: number) {
+        this.ctx.beginPath();
+        this.ctx.arc(x, y, r, sa, ea);
+        this.ctx.stroke(); 
+    }
+
+    /**
+     * Fills a polygon at the given coorinates with the radius and number of sides specified.
```
Comment (anbclausen) [inline graphics.ts]: coordinates*

File: graphics.ts [hunk 6]
```diff
@@ -97,8 +171,159 @@ class Graphics {
...
+     * @param r The radius of the polygon.
+     * @param n The number of sides in the polygon
+     * @param v The rotation of the polygon in radians. Default is 0.
+     */
+    fillPolygon(x: number, y: number, r: number, n: number, v: number = 0) {
+        this.makePolygon(x, y, r, n, true, v);
+    }
+
+    /**
+     * Draws a polygon at the given coorinates with the radius and number of sides specified.
```
Comment (anbclausen) [inline graphics.ts]: coordinates

File: graphics.ts [hunk 7]
```diff
@@ -97,8 +171,159 @@ class Graphics {
...
+     * @param x The x coordinate for the center of the polygon.
+     * @param y The y coordinate for the center of the polygon.
+     * @param r The radius of the polygon.
+     * @param n The number of sides in the polygon
+     * @param v The rotation of the polygon in radians. Default is 0.
+     */
+    drawPolygon(x: number, y: number, r: number, n: number, v: number = 0) {
```
Comment (anbclausen) [inline graphics.ts]: again, i don't like default values

File: graphics.ts [hunk 8]
```diff
@@ -97,8 +171,159 @@ class Graphics {
...
+    /**
+     * The actual implementation of the making of the polygon. Draws the polygon, and fills it if specified.
+     * @param x The x coordinate for the center of the polygon.
+     * @param y The y coordinate for the center of the polygon.
+     * @param r The radius of the polygon.
+     * @param n The number of sides in the polygon
+     * @param fill A boolean indicating if the polygon should be filled.
+     * @param v The rotation of the polygon in radians.
+     */
+    private makePolygon(x: number, y: number, r: number, n: number, fill: boolean, v: number) {
```
Comment (anbclausen) [inline graphics.ts]: Great refactoring, creating this method. I think you should refactor it to use already existing methods such as the translate and rotate methods you created earlier :)

File: graphics.ts [hunk 9]
```diff
@@ -97,8 +171,159 @@ class Graphics {
...
+
+    /**
+     * The actial implementation that makes the custom shape, and fills it if specified.
+     * @param ps The array containing tuples of coordinate points.
+     * @param fill A boolean indicating if the shape should be filled.
+     */
+    private makeShape(ps: [number, number][], fill: boolean) {
+        this.ctx.beginPath();
+        this.ctx.moveTo(ps[0][0], ps[0][1]);
+        ps.forEach(p => {
```
Comment (anbclausen) [inline graphics.ts:321]: for some reason it bugs me, that you use .forEach here, and for in makePolygon.. :D

**Review by anbclausen — CHANGES_REQUESTED** [review]
Great ideas, please check this review!

File: graphics.ts (hunk 9 above)
Comment (johandegn) [reply graphics.ts:321]: Yea, i prefer the simplicity of the forEach but it is only in makeShape i operate on an Array. I'll leave it like this for now, unless you specifically request it to match ;)

**Comment by johandegn** [comment]
Okay, it should now all be fixed.

One note: i also created two simple methods to convert degrees and radians.

**Comment by johandegn** [comment]
Ups, accidentally closed request. Reopened.

**Comment by johandegn** [comment]
It should now all be fixed.

File: graphics.ts (hunk 9 above)
Comment (anbclausen) [reply graphics.ts:321]: Fair point. Leave it like this :)

**Review by anbclausen — APPROVED** [review]
Great functionality and beautiful code! Thanks!

**Pull request merged**